smartsim/_core/control/previewrenderer.py

```diff
@@ -0,0 +1,198 @@
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import pathlib
+import typing as t
+from enum import Enum
+
+import jinja2
+import jinja2.utils as u
+from jinja2 import pass_eval_context
+
+from ..._core.config import CONFIG
+from ..._core.control import Manifest
+from ...error.errors import PreviewFormatError
+from ...log import get_logger
+from .job import Job
+
+logger = get_logger(__name__)
+
+if t.TYPE_CHECKING:
+    from smartsim import Experiment
+
+
+class Format(str, Enum):
+    PLAINTEXT = "plain_text"
+
+
+class Verbosity(str, Enum):
+    INFO = "info"
+    DEBUG = "debug"
+    DEVELOPER = "developer"
+
+
+@pass_eval_context
+def as_toggle(_eval_ctx: u.F, value: bool) -> str:
+    """Return "On" if value returns True,
+    and "Off" is value returns False.
+    """
+    return "On" if value else "Off"
+
+
+@pass_eval_context
+def get_ifname(_eval_ctx: u.F, value: t.List[str]) -> str:
+    """Extract Network Interface from orchestrator run settings."""
+    if value:
+        for val in value:
+            if "ifname=" in val:
+                output = val.split("=")[-1]
+                return output
+    return ""
+
+
+@pass_eval_context
+def get_dbtype(_eval_ctx: u.F, value: str) -> str:
+    """Extract data base type."""
+    if value:
+        if "-cli" in value:
+            db_type, _ = value.split("/")[-1].split("-", 1)
+            return db_type
+    return ""
+
+
+@pass_eval_context
+def is_list(_eval_ctx: u.F, value: str) -> bool:
+    """Return True if item is of type list, and False
+    otherwise, to determine how Jinja template should
+    render an item.
+    """
+    return isinstance(value, list)
+
+
+def render_to_file(content: str, filename: str) -> None:
+    """Output preview to a file if an output filename
+    is specified.
+
+    :param content: The rendered preview.
+    :type content: str
```

Comment:
I think we can go through and remove all of our `:type xxx:` hints in the docstrings now that they're auto-generated.

Reply:
removed.